### PR TITLE
Add braille text wrap modes with continuation marks

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Part of #17010. Split out from #19916 at reviewer request — this is the second of three PRs. The pyphen abstraction layer was shipped in #20145; the syllable-aware wrap mode follows in a separate PR.

### Summary of the issue:
The braille word-wrap setting is a single boolean that gives users no control over how words are broken at the display edge. When a word is cut mid-way there is also no visual indication that the word continues on the next row.

### Description of user facing changes:
The boolean "word wrap" checkbox in the braille settings has been replaced with a **Text wrap** combo box with three choices:

* **Off** — Wrap at the raw edge of the display, cutting words in the middle if necessary. No visual indication that a word was cut.
* **Show mark when words are cut** — Wrap at the raw edge, but whenever a word is cut mid-way, replace the last cell of the row with a continuation mark (braille dots 7-8) so the reader knows the word continues on the next row.
* **At word boundaries** — Prefer breaking at spaces. If no space fits on the row, fall back to cutting the word and showing the continuation mark.

Existing config profiles with the old setting are automatically upgraded.

### Description of developer facing changes:
* `BrailleTextWrapFlag` feature flag enum added to `config.featureFlagEnums` with members `DEFAULT`, `NONE`, `MARK_WORD_CUTS`, `AT_WORD_BOUNDARIES`.
* Config schema bumped v22 → v23; old `wordWrap` boolean is deprecated and bridged bidirectionally to `textWrap` via `_linkDeprecatedValues`, so add-ons reading or writing the old key keep working (with a deprecation warning).
* `CONTINUATION_SHAPE = 0xC0` (dots 7-8) constant added to `braille`.
* `_WindowRowPositions` frozen dataclass added to `braille` to hold the start/end buffer positions and continuation-mark flag for each row of the braille window, replacing the previous anonymous tuple.

### Description of development approach:
The continuation mark is unified: it consistently means "a word was cut here" regardless of mode, so readers get a predictable signal.

`BrailleBuffer._calculateWindowRowBufferOffsets` is extended to implement all three modes. Each entry in `_windowRowBufferOffsets` is a `_WindowRowPositions` instance whose `showContinuationMark` field records whether that row needs a continuation mark. `BrailleBuffer._get_windowBrailleCells` reads that flag to insert the mark. `BrailleBuffer._set_windowEndPos` short-circuits space-seeking for `NONE` and `MARK_WORD_CUTS` modes (backwards scroll alignment).

### Testing strategy:
Unit tests cover:
* All three wrap modes and their edge cases in `_calculateWindowRowBufferOffsets` (no cells, fits exactly, word cut, word too long for display, no space on line).
* Continuation-mark rendering in `_get_windowBrailleCells`.
* Profile upgrade from `wordWrap = True`/`False` → correct `textWrap` flag values.

Manual testing: loaded a pre-upgrade profile with `wordWrap = True`/`False` and confirmed the upgrade writes the expected `textWrap` value and the braille settings panel shows the matching label.

### Known issues with pull request:
None.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/config/profileUpgradeSteps.py [hunk 1]
```diff
@@ -687,3 +688,28 @@ def upgradeConfigFrom_21_to_22(profile: ConfigObj):
 	if language.casefold() == "auto":
 		speechConf["language"] = "en"
 		log.debug("Changed math.speech.language from 'Auto' to 'en'.")
+
+
+def upgradeConfigFrom_22_to_23(profile: ConfigObj) -> None:
+	"""
+	If the wordWrap braille config flag is explicitly set in a profile,
+	set the new text wrap option to word boundaries,
+	rather than the new default of word or syllable boundaries.
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/config/profileUpgradeSteps.py]: Fixed. Changed "word or syllable boundaries" to "at word boundaries" in the docstring.

File: source/braille.py [hunk 2]
```diff
@@ -326,6 +330,7 @@
 	(0xFF, _("All dots")),
 )
 SELECTION_SHAPE = 0xC0  #: Dots 7 and 8
+CONTINUATION_SHAPE = 0xC0  #: Dots 7 and 8
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py:334]: Intentionally kept separate. These are conceptually distinct (continuation marker vs. selection indicator) that happen to share the same dot pattern today. Aliasing them would couple two independent concerns — a future change to either pattern would become confusing. The inline comments already document the dot value.

File: source/braille.py [hunk 3]
```diff
@@ -1954,36 +1987,48 @@ def _get_windowStartPos(self) -> int:
 	def _set_windowStartPos(self, pos: int) -> None:
 		self._calculateWindowRowBufferOffsets(pos)
 
+	def _isMidWordCut(self, end: int, bufferEnd: int) -> bool:
+		"""Return True when the cut at `end` falls in the middle of a word (both adjacent cells are non-space)."""
+		return end < bufferEnd and all(self.brailleCells[end - 1 : end + 1])
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py:2005]: Not a real edge case. Every callsite computes `end = start + numCols` where `numCols >= 1`, so `end >= 1` is structurally guaranteed. Adding a guard for an impossible input would be dead code.

File: source/config/featureFlagEnums.py [hunk 4]
```diff
@@ -139,6 +139,30 @@ def _displayStringLabels(self) -> dict["FontFormattingBrailleModeFlag", str]:
...
+			# Translators: A choice in a combo box in the braille settings panel to configure text wrapping.
+			self.MARK_WORD_CUTS: pgettext("braille text wrap", "Show mark when words are cut"),
+			# Translators: A choice in a combo box in the braille settings panel to configure text wrapping.
+			self.AT_WORD_BOUNDARIES: pgettext("braille text wrap", "At word boundaries"),
+		}
+
+	DEFAULT = enum.auto()
+	NONE = enum.auto()
+	MARK_WORD_CUTS = enum.auto()
+	AT_WORD_BOUNDARIES = enum.auto()
```
Comment (seanbudd) [inline source/config/featureFlagEnums.py]: these should go at the top of the enum

File: source/gui/settingsDialogs.py [hunk 5]
```diff
@@ -5499,12 +5499,15 @@ def makeSettings(self, settingsSizer):
 			list(braille.BrailleMode)[self.brailleModes.GetSelection()] is braille.BrailleMode.FOLLOW_CURSORS,
 		)
 
-		# Translators: The label for a setting in braille settings to enable word wrap
-		# (try to avoid splitting words at the end of the braille display).
-		wordWrapText = _("Avoid splitting &words when possible")
-		self.wordWrapCheckBox = sHelper.addItem(wx.CheckBox(self, label=wordWrapText))
-		self.bindHelpEvent("BrailleSettingsWordWrap", self.wordWrapCheckBox)
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:5506]: I think we should possibly retain the same anchor

File: source/braille.py [hunk 6]
```diff
@@ -1826,10 +1838,10 @@ def __init__(self, handler):
 		#: The translated braille representation of the entire buffer.
 		#: @type: [int, ...]
 		self.brailleCells = []
-		self._windowRowBufferOffsets: list[tuple[int, int]] = [(0, 0)]
+		self._windowRowBufferOffsets: list[tuple[int, int, bool]] = [(0, 0, False)]
```
Comment (seanbudd) [inline source/braille.py]: can we turn this into a dataclass please?

Comment (LeonarddeR) [reply source/braille.py]: Makes sense, done.

File: source/braille.py [hunk 7]
```diff
@@ -1954,36 +2000,48 @@ def _get_windowStartPos(self) -> int:
...
+						# lastSpaceIndex < end proves brailleCells[end] is non-zero,
+						# so searching [start, end) yields the same lastSpaceIndex.
+						end = lastSpaceIndex + 1
 				except (ValueError, IndexError):
-					pass  # No space on line
-			self._windowRowBufferOffsets.append((start, end))
+					# No space on line - fall back to display-edge cut.
+					if self._isMidWordCut(end, bufferEnd):
+						end -= 1
+						showContinuationMark = True
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py:2043]: This is existing logic I don't dare to change honestly.

File: user_docs/en/userGuide.md [hunk 8]
```diff
@@ -2584,18 +2584,22 @@ Automatic scrolling will be disabled if a routing key is pressed, if a message i
...
+The following options are available:
+
+* Off: Text wraps at the display edge, cutting words mid-way if necessary, without showing any continuation mark.
+As much of the text as possible will be displayed on each row.
+When you scroll the display, you will be able to read the rest of the text.
+* Show mark when words are cut: Text is not wrapped, but whenever a word is cut at the end of the display, a continuation mark is shown.
+When you scroll the display, you will be able to read the rest of the word.
+* At word boundaries: A word which is too large to fit at the end of the braille display will not be split.
 Instead, there will be some blank space at the end of the display.
 When you scroll the display, you will be able to read the entire word.
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply user_docs/en/userGuide.md:2600]: I think this is a false positive. Common pattern in changes.md as well.

File: source/config/profileUpgradeSteps.py [hunk 9]
```diff
@@ -687,3 +688,28 @@ def upgradeConfigFrom_21_to_22(profile: ConfigObj):
...
+		return
+	except ValueError:
+		log.error(f"'{key}' is not a boolean, got {profile[section][key]!r}. No action taken.")
+		return
+
+	newValue = BrailleTextWrapFlag.AT_WORD_BOUNDARIES.name if oldValue else BrailleTextWrapFlag.NONE.name
+	profile[section][newKey] = newValue
+	log.debug(
+		f"Converted '{key}' with value {oldValue} to '{newKey}' with value {newValue}. Deleted '{key}'.",
+	)
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/config/profileUpgradeSteps.py:715]: Fixed. Note, wordWrap shouldn't be deleted from the profile as it is still a linked value, so whenever it is changed using text wrap, it comes back anyway.

**Review by seanbudd — APPROVED** [review]

File: user_docs/en/userGuide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (seanbudd) [reply user_docs/en/userGuide.md:2600]: yes copilot has been incorrectly raising this a lot lately

**Review by Qchristensen — APPROVED** [review]
Reads well

**Comment by CyrilleB79** [comment]
Hi @LeonarddeR,

Just translating NVDA and discovering this feature a bit more.

It's quite strange to read "Show mark when words are cut" as a value for "Text wrap". The other values (Off, At word boundaries, At word or syllable boundaries) indicate in which condition the text wraps. On the opposite, "Show mark when words are cut" is not a condition that triggers text wrapping. Why not having used only 3 values added a separate checkbox to choose if marks are shown or not; if this only applies when Off is selected, the checkbox can be greyed out for other values.

Alternatively, if you find this suggestion irrelevant or too late in 2026.3 dev cycle, maybe we can replace "Show mark when words are cut" by "Off (Show mark when words are cut)" or "Off, show mark when words are cut", so that the value remains a trigger condition for text wrap.

**Comment by LeonarddeR** [comment]
@CyrilleB79 wrote:

> Why not having used only 3 values added a separate checkbox to choose if marks are shown or not; if this only applies when Off is selected, the checkbox can be greyed out for other values.

I have considered the UX but it wasn't easy to get it right. The problem with your suggestion is that syllable boundaries implicitly show marks, whereas it doesn't make sense to show marks for word boundaries at all because the word then ends just fine. 

> Alternatively, if you find this suggestion irrelevant or too late in 2026.3 dev cycle, 

That's not up to me 😉

> maybe we can replace "Show mark when words are cut" by "Off (Show mark when words are cut)" or "Off, show mark when words are cut", so that the value remains a trigger condition for text wrap.

I think all depends on how your define text wrap. 
If I understand it correctly from the CSS side of things, text either wraps or overflows.
See for example the [documentation of the overflow-wrap property](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/overflow-wrap).
Thinking about it from these lines, you can even defend that text wrap being off in braille is actually a non-existing thing. Text wrap being off definitely doesn't mean overflow, as that would be extremely weird.
So even when text wrap is off, in Braille text wraps when there is no more space on the display, there's just no smart wrapping.
So from another angle, we can also consider changing `text wrap` to `smart text wrap` or `overflow wrap` or something.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 wrote:
> 
> > Why not having used only 3 values added a separate checkbox to choose if marks are shown or not; if this only applies when Off is selected, the checkbox can be greyed out for other values.
> 
> I have considered the UX but it wasn't easy to get it right. The problem with your suggestion is that syllable boundaries implicitly show marks, whereas it doesn't make sense to show marks for word boundaries at all because the word then ends just fine.

Yes this sounds logical to me too thinking a bit more...


> > maybe we can replace "Show mark when words are cut" by "Off (Show mark when words are cut)" or "Off, show mark when words are cut", so that the value remains a trigger condition for text wrap.
> 
> I think all depends on how your define text wrap. If I understand it correctly from the CSS side of things, text either wraps or overflows. See for example the [documentation of the overflow-wrap property](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/overflow-wrap). Thinking about it from these lines, you can even defend that text wrap being off in braille is actually a non-existing thing. Text wrap being off definitely doesn't mean overflow, as that would be extremely weird. So even when text wrap is off, in Braille text wraps when there is no more space on the display, there's just no smart wrapping. So from another angle, we can also consider changing `text wrap` to `smart text wrap` or `overflow wrap` or something.

That's true, the text always wraps, fortunately, else it would just be cropped / truncated... So "Text wrap" = "Off" is not correct either.

I'm not a fan of "Smart text wrap": at the end we never know why something is smart and another not. "smart" is not very descriptive, and thus, it becomes also harder to translate.

I've a new suggestion: keep the "Text wrap" option name with the following values:
* "Anywhere" (replacing Off)
* "Anywhere, with mark when words are cut" (replacing "Show mark when words are cut")
* "At word boundaries" (unchanged)
* "At word or syllable boundaries" (unchanged)
You may want to reword it still bettter but the idea is there. What do you think?

**Comment by LeonarddeR** [comment]
@CyrilleB79 I think I like it. Can you file it in a new issue?
